### PR TITLE
util: add `CloneBoxService`

### PR DESCRIPTION
**Pull request by davidpdrsn**
This upstreams a little utility I'm using a bunch in axum. Its often
useful to erase the type of a service while still being able to clone
it.

`BoxService` isn't `Clone` so previously you had to combine it with
`Buffer` but doing that a lot (which we did in axum) had measurable
impact on performance.

Note this doesn't add a `!Send` variant. I personally have no use
for that but if someone thinks it would be nice I can do a follow up PR.

### CODE REVIEW

File: tower/src/util/clone_boxed.rs
Comment (hawkw) [inline tower/src/util/clone_boxed.rs:10]: nit, TIOLI: should we maybe link to the docs for the `Clone` and `Service` traits?

File: tower/src/util/clone_boxed.rs
Comment (hawkw) [inline tower/src/util/clone_boxed.rs:13]: it would be nice if the docs for this included a comparison with the non-`Clone` `BoxService` type? something like "This is similar to the [`BoxService`] type, but ..."

File: tower/src/util/clone_boxed.rs
Comment (hawkw) [inline tower/src/util/clone_boxed.rs:45]: minor nit: when used as a contraction for "it is", you want "it's", rather than "its":
```suggestion
/// // `CloneService` will erase the type so it's nameable
```

File: tower/src/util/clone_boxed.rs
Comment (hawkw) [inline tower/src/util/clone_boxed.rs:134]: the debug output doesn't include any of these type parameters, so I believe these can all be removed, and the type can always be `Debug`:
```suggestion
impl<T, U, E> fmt::Debug for CloneBoxService<T, U, E> {
```

File: tower/src/util/clone_boxed.rs
Comment (hawkw) [inline tower/src/util/clone_boxed.rs:136]: take it or leave it: we _may_ want to consider including the request and response type _names_ here, so different instances of `CloneBoxService` can be distinguished? Maybe something like
```suggestion
        write!(
            fmt, 
            "CloneBoxService<Request = {}, Response = {}, Error = {}>",
            std::any::type_name::<T>(),
            std::any::type_name::<U>(),
            std::any::type_name::<E>(),
        )
```
but, on the other hand, this could be quite long, and we might want to correctly handle multi-line formatting in alt mode (`{:#}`)...so, maybe that's something to hold off on and add later if it's needed...

Comment (hawkw) [reply tower/src/util/clone_boxed.rs:136]: also, `BoxService` doesn't do this, so it's probably better left for a separate PR if we _do_ want to add something like that...

**Review by hawkw — CHANGES_REQUESTED** [review]
Overall, this looks great to me --- we may also want to add methods to `ServiceExt` and/or `ServiceBuilder` for wrapping a service in `CloneBoxService`, but we can do that in a follow up PR.

I commented on a few minor things. The only one that I think is actually important to address before merging is the unnecessary `fmt::Debug` bounds on `CloneBoxService`'s type parameters: since those types are never formatted in the `Debug` impl, it shouldn't be necessary for them to implement `Debug`.

Beyond that, this looks good to me and I'll be happy to approve it once that's addressed!

File: tower/src/util/clone_boxed.rs
Comment (davidpdrsn) [reply tower/src/util/clone_boxed.rs:45]: One day I'll learn this 😜

File: tower/src/util/clone_boxed.rs
Comment (davidpdrsn) [reply tower/src/util/clone_boxed.rs:136]: I was actually thinking the type params are there such that this could be improved in the future, without a breaking change. I think I'll leave it as it is now and make an issue about improving the formatting, for `BoxService` as well.

File: tower/CHANGELOG.md
Comment (olix0r) [inline tower/CHANGELOG.md:10]: Here and elsewhere in the comments you refer to `CloneService`, but I think you mean `CloneBoxService`?
```suggestion
- **util**: Add `CloneBoxService` which is a `Clone + Send` boxed `Service`.
```

File: tower/src/util/clone_boxed.rs
Comment (olix0r) [reply tower/src/util/clone_boxed.rs:45]:
```suggestion
/// // `CloneBoxService` will erase the type so it's nameable.
```

File: tower/CHANGELOG.md
Comment (davidpdrsn) [reply tower/CHANGELOG.md]: Ah yes good catch! I was a bit too quick on the auto complete.

File: tower/src/util/clone_boxed.rs
Comment (hawkw) [reply tower/src/util/clone_boxed.rs:139]: > I was actually thinking the type params are there such that this could be improved in the future,

I don't think we want the type params to be `Debug` in any case; we can't actually include values of those types in the formatted output, since they're not part of the struct. I think the where clause won't ever be useful for future improvements to the debug format, since we'll never have instances of those types to format when formatting a `CloneBoxService` instance. We can use `type_name` to get the _names_ of those types without requiring those types to be `Debug`.

File: tower/src/util/clone_boxed.rs
Comment (davidpdrsn) [reply tower/src/util/clone_boxed.rs]: I fixed the typos in https://github.com/tower-rs/tower/pull/615/commits/28915aa9d1cab86aebb38a8fd6f928f0bd1c58d4

**Comment by davidpdrsn** [comment]
@hawkw I think adding more convenience methods to `ServiceExt` and `ServiceBuilder` for this is a good idea. I'll do that in a follow up PR.

File: tower/src/util/clone_boxed.rs
Comment (davidpdrsn) [reply tower/src/util/clone_boxed.rs:134]: Oh yeah doh of course! I've removed the bounds.

**Review by hawkw — APPROVED** [review]
okay, this looks good to me! thanks!

**Pull request merged**
